File: src/aws/lambdas/incremental_distribution/src/main/java/uk/nhs/nhsx/keyfederation/download/ExposureDownload.java
package uk.nhs.nhsx.keyfederation.download;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;

import java.util.List;

public class ExposureDownload {
    
    public final String keyData;
    public final int rollingStartNumber;
    public final int transmissionRiskLevel;
    public final int rollingPeriod;
    public final String origin;
    public final List<String> regions;

    @JsonCreator
    public ExposureDownload(@JsonProperty String keyData,
                            @JsonProperty int rollingStartNumber,
                            @JsonProperty int transmissionRiskLevel,
                            @JsonProperty int rollingPeriod,
                            @JsonProperty String origin,
                            @JsonProperty List<String> regions) {
        this.keyData = keyData;
        this.rollingStartNumber = rollingStartNumber;
        this.transmissionRiskLevel = transmissionRiskLevel;
        this.rollingPeriod = rollingPeriod;
        this.origin = origin;
        this.regions = regions;
    }
}

File: src/aws/lambdas/incremental_distribution/src/main/java/uk/nhs/nhsx/keyfederation/download/ExposureDownload.kt
package uk.nhs.nhsx.keyfederation.download

data class ExposureDownload(val keyData: String,
                            val rollingStartNumber: Int,
                            val transmissionRiskLevel: Int,
                            val rollingPeriod: Int,
                            val origin: String,
                            val regions: List<String>,
                            val testType: TestType,
                            val reportType: ReportType,
                            val daysSinceOnset: Int)

enum class TestType(val wireValue: Int) {
    PCR(1),
    LFT(2),
    SELF_ASSISTED_LFT(3);

    override fun toString() = wireValue.toString()

    companion object {
        fun from(wire: Int) =
            values().first { it.wireValue == wire }
    }
}

enum class ReportType(val wireValue: Int) {
    UNKNOWN(0),
    CONFIRMED_TEST(1),
    CONFIRMED_CLINICAL_DIAGNOSIS(2),
    SELF_REPORT(3),
    RECURSIVE(4),
    REVOKED(5);

    override fun toString() = wireValue.toString()

    companion object {
        fun from(wire: Int) =
            values().first { it.wireValue == wire }
    }
}
